File: app/routers/finance.py
from datetime import datetime, date
from fastapi import APIRouter, Request, Depends, Form, HTTPException
from fastapi.responses import HTMLResponse, RedirectResponse
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import func, extract
from typing import Optional

from ..database import get_db
from ..models import Expense, ExpenseCategory, Payment, Transaction

router = APIRouter(tags=["finance"])


def require_auth(request: Request):
    if not request.session.get("user_id"):
        return RedirectResponse(url="/login", status_code=302)
    return None


def require_admin(request: Request):
    if not request.session.get("user_id"):
        return RedirectResponse(url="/login", status_code=302)
    if request.session.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Chỉ admin mới có quyền thực hiện thao tác này")
    return None


# ─────────────────────────────────────────────
# GET /finance  →  trang tổng quan thu chi
# ─────────────────────────────────────────────

@router.get("/finance", response_class=HTMLResponse)
async def finance_index(
    request: Request,
    month: Optional[str] = None,
    db: Session = Depends(get_db),
):
    redirect = require_auth(request)
    if redirect:
        return redirect

    templates = request.app.state.templates
    today = date.today()

    # Parse month filter (YYYY-MM)
    if month:
        try:
            year, mon = map(int, month.split("-"))
        except ValueError:
            year, mon = today.year, today.month
    else:
        year, mon = today.year, today.month
        month = today.strftime("%Y-%m")

    # Tổng thu hôm nay (từ payments của completed orders)
    total_income_today = (
        db.query(func.coalesce(func.sum(Payment.amount), 0))
        .filter(func.date(Payment.paid_at) == today)
        .scalar()
    ) or 0.0

    # Tổng chi hôm nay
    total_expense_today = (
        db.query(func.coalesce(func.sum(Expense.amount), 0))
        .filter(func.date(Expense.occurred_at) == today)
        .scalar()
    ) or 0.0

    profit_today = total_income_today - total_expense_today

    # Danh sách chi phí trong tháng được chọn
    expenses = (
        db.query(Expense)
        .filter(
            extract("year", Expense.occurred_at) == year,
            extract("month", Expense.occurred_at) == mon,
        )
        .options(joinedload(Expense.category), joinedload(Expense.user))
        .order_by(Expense.occurred_at.desc())
        .all()
    )

    # Danh mục chi
    categories = (
        db.query(ExpenseCategory)
        .filter(ExpenseCategory.is_active == True)
        .order_by(ExpenseCategory.name)
        .all()
    )

    return templates.TemplateResponse(
        "finance/index.html",
        {
            "request": request,
            "total_income_today": total_income_today,
            "total_expense_today": total_expense_today,
            "profit_today": profit_today,
            "expenses": expenses,
            "categories": categories,
            "month": month,
            "today": today.isoformat(),
        },
    )


# ─────────────────────────────────────────────
# POST /finance/expenses  →  tạo khoản chi mới
# ─────────────────────────────────────────────

@router.post("/finance/expenses")
async def create_expense(
    request: Request,
    category_id: int = Form(...),
    amount: float = Form(...),
    description: str = Form(...),
    occurred_at: str = Form(...),
    db: Session = Depends(get_db),
):
    redirect = require_auth(request)
    if redirect:
        return redirect

    user_id = request.session.get("user_id")

    try:
        occurred_dt = datetime.fromisoformat(occurred_at)
    except ValueError:
        occurred_dt = datetime.utcnow()

    expense = Expense(
        category_id=category_id,
        user_id=user_id,
        amount=amount,
        description=description,
        occurred_at=occurred_dt,
    )
    db.add(expense)

    # Tạo Transaction record cho khoản chi
    transaction = Transaction(
        type="expense",
        amount=amount,
        description=description,
        reference_type="expense",
        user_id=user_id,
        occurred_at=occurred_dt,
    )
    db.add(transaction)
    db.commit()

    flash = request.session.setdefault("flash", [])
    flash.append({"type": "success", "message": "Đã ghi nhận khoản chi"})

    return RedirectResponse(url="/finance", status_code=302)


# ─────────────────────────────────────────────
# GET /finance/expenses/{id}/edit  →  form sửa
# ─────────────────────────────────────────────

@router.get("/finance/expenses/{expense_id}/edit", response_class=HTMLResponse)
async def edit_expense_form(
    request: Request,
    expense_id: int,
    db: Session = Depends(get_db),
):
    redirect = require_auth(request)
    if redirect:
        return redirect

    templates = request.app.state.templates

    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Không tìm thấy khoản chi")

    categories = (
        db.query(ExpenseCategory)
        .filter(ExpenseCategory.is_active == True)
        .order_by(ExpenseCategory.name)
        .all()
    )

    return templates.TemplateResponse(
        "finance/edit_expense.html",
        {
            "request": request,
            "expense": expense,
            "categories": categories,
        },
    )


# ─────────────────────────────────────────────
# POST /finance/expenses/{id}/edit  →  lưu sửa
# ─────────────────────────────────────────────

@router.post("/finance/expenses/{expense_id}/edit")
async def update_expense(
    request: Request,
    expense_id: int,
    category_id: int = Form(...),
    amount: float = Form(...),
    description: str = Form(...),
    occurred_at: str = Form(...),
    db: Session = Depends(get_db),
):
    err = require_admin(request)
    if err:
        return err

    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Không tìm thấy khoản chi")

    try:
        occurred_dt = datetime.fromisoformat(occurred_at)
    except ValueError:
        occurred_dt = expense.occurred_at

    expense.category_id = category_id
    expense.amount = amount
    expense.description = description
    expense.occurred_at = occurred_dt
    expense.updated_at = datetime.utcnow()

    db.commit()

    flash = request.session.setdefault("flash", [])
    flash.append({"type": "success", "message": "Đã cập nhật khoản chi"})

    return RedirectResponse(url="/finance", status_code=302)


# ─────────────────────────────────────────────
# POST /finance/expenses/{id}/delete  →  xóa
# ─────────────────────────────────────────────

@router.post("/finance/expenses/{expense_id}/delete")
async def delete_expense(
    request: Request,
    expense_id: int,
    db: Session = Depends(get_db),
):
    err = require_admin(request)
    if err:
        return err

    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Không tìm thấy khoản chi")

    db.delete(expense)
    db.commit()

    flash = request.session.setdefault("flash", [])
    flash.append({"type": "success", "message": "Đã xóa khoản chi"})

    return RedirectResponse(url="/finance", status_code=302)


# ─────────────────────────────────────────────
# GET /finance/categories  →  quản lý danh mục
# ─────────────────────────────────────────────

@router.get("/finance/categories", response_class=HTMLResponse)
async def categories_page(
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_auth(request)
    if redirect:
        return redirect

    templates = request.app.state.templates

    categories = (
        db.query(ExpenseCategory)
        .order_by(ExpenseCategory.name)
        .all()
    )

    return templates.TemplateResponse(
        "finance/categories.html",
        {
            "request": request,
            "categories": categories,
        },
    )


# ─────────────────────────────────────────────
# POST /finance/categories  →  tạo danh mục
# ─────────────────────────────────────────────

@router.post("/finance/categories")
async def create_category(
    request: Request,
    name: str = Form(...),
    description: str = Form(default=""),
    db: Session = Depends(get_db),
):
    redirect = require_auth(request)
    if redirect:
        return redirect

    cat = ExpenseCategory(
        name=name,
        description=description if description else None,
    )
    db.add(cat)
    db.commit()

    flash = request.session.setdefault("flash", [])
    flash.append({"type": "success", "message": f"Đã tạo danh mục '{name}'"})

    return RedirectResponse(url="/finance/categories", status_code=302)


# ─────────────────────────────────────────────
# POST /finance/categories/{id}/delete  →  xóa danh mục
# ─────────────────────────────────────────────

@router.post("/finance/categories/{cat_id}/delete")
async def delete_category(
    request: Request,
    cat_id: int,
    db: Session = Depends(get_db),
):
    err = require_admin(request)
    if err:
        return err

    cat = db.query(ExpenseCategory).filter(ExpenseCategory.id == cat_id).first()
    if not cat:
        raise HTTPException(status_code=404, detail="Không tìm thấy danh mục")

    # Kiểm tra có expense đang dùng không
    count = db.query(Expense).filter(Expense.category_id == cat_id).count()
    if count > 0:
        flash = request.session.setdefault("flash", [])
        flash.append({"type": "error", "message": f"Không thể xóa danh mục đang có {count} khoản chi"})
        return RedirectResponse(url="/finance/categories", status_code=302)

    db.delete(cat)
    db.commit()

    flash = request.session.setdefault("flash", [])
    flash.append({"type": "success", "message": "Đã xóa danh mục"})

    return RedirectResponse(url="/finance/categories", status_code=302)

File: src/routes/finance.js
const express = require('express');
const router = express.Router();
const { q } = require('../db');
const { requireAuth, requireAdmin } = require('../middleware/auth');

// LƯU Ý: trước đây có 3 route admin sửa/xoá khoản chi
// (GET/POST /expenses/:id/edit, POST /expenses/:id/delete) — đã gỡ theo
// quyết định nghiệp vụ: khoản chi đã ghi nhận là immutable, sai thì
// ghi nhận khoản đối ứng mới chứ không sửa lịch sử. Nếu cần escape
// hatch, làm trực tiếp DB qua docker exec.

// ─────────────────────────────────────────────
// GET /finance — tổng quan thu chi
// ─────────────────────────────────────────────
router.get('/', requireAuth, (req, res) => {
  // Doanh thu hôm nay (từ payments của orders đã hoàn thành)
  const incomeRow = q.get(
    `SELECT COALESCE(SUM(p.amount), 0) AS total
     FROM payments p
     JOIN orders o ON o.id = p.order_id
     WHERE o.status = 'completed'
       AND date(p.paid_at) = date('now', 'localtime')`
  );
  const todayIncome = incomeRow ? incomeRow.total : 0;

  // Chi phí hôm nay
  const expenseRow = q.get(
    `SELECT COALESCE(SUM(amount), 0) AS total
     FROM expenses
     WHERE date(occurred_at) = date('now', 'localtime')`
  );
  const todayExpense = expenseRow ? expenseRow.total : 0;

  const profit = todayIncome - todayExpense;

  // Tháng lọc (mặc định tháng hiện tại)
  const now = new Date();
  const defaultMonth = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}`;
  const selectedMonth = req.query.month || defaultMonth;

  // Chi phí theo tháng đã chọn (kèm category_name + user full_name)
  const expenses = q.all(
    `SELECT e.*, ec.name AS category_name, u.full_name AS user_full_name
     FROM expenses e
     JOIN expense_categories ec ON ec.id = e.category_id
     LEFT JOIN users u ON u.id = e.user_id
     WHERE strftime('%Y-%m', e.occurred_at) = ?
     ORDER BY e.occurred_at DESC`,
    selectedMonth
  );

  // Khoản thu theo tháng — từng payment của các order completed.
  // Dùng paid_at thay vì orders.updated_at để không bị nhảy ngày khi
  // hoá đơn được sửa (gặp gotcha với ORD-20260513-023).
  const incomes = q.all(
    `SELECT p.id AS payment_id, p.amount, p.paid_at,
            pm.name AS method_name,
            o.id AS order_id, o.order_code,
            t.name AS table_name
     FROM payments p
     JOIN orders o ON o.id = p.order_id
     JOIN tables t ON t.id = o.table_id
     JOIN payment_methods pm ON pm.id = p.method_id
     WHERE o.status = 'completed'
       AND strftime('%Y-%m', p.paid_at) = ?
     ORDER BY p.paid_at DESC`,
    selectedMonth
  );

  // Danh mục chi phí đang hoạt động
  const categories = q.all(
    `SELECT * FROM expense_categories WHERE is_active = 1 ORDER BY name`
  );

  const total_expense = expenses.reduce((sum, e) => sum + (e.amount || 0), 0);
  const total_income  = incomes.reduce((sum, i) => sum + (i.amount  || 0), 0);

  res.render('finance/index.html', {
    todayIncome,
    todayExpense,
    profit,
    expenses,
    incomes,
    categories,
    selectedMonth,
    total_expense,
    total_income,
    // Aliases cho template cũ (nếu template dùng tên khác)
    total_income_today:   todayIncome,
    total_expense_today:  todayExpense,
    profit_today:         profit,
    month:                selectedMonth,
  });
});

// ─────────────────────────────────────────────
// POST /finance/expenses — thêm khoản chi
// ─────────────────────────────────────────────
router.post('/expenses', requireAuth, (req, res) => {
  const categoryId  = parseInt(req.body.category_id);
  const amount      = parseFloat(req.body.amount);
  const description = (req.body.description || '').trim();
  const occurredAt  = req.body.occurred_at || new Date().toISOString().slice(0, 16);

  if (!categoryId || !amount || amount <= 0 || !description) {
    res.flash('error', 'Vui lòng điền đầy đủ thông tin khoản chi.');
    return res.redirect('/finance');
  }

  try {
    const addExpense = q.transaction(() => {
      // Ghi expense
      const result = q.run(
        `INSERT INTO expenses (category_id, user_id, amount, description, occurred_at)
         VALUES (?, ?, ?, ?, ?)`,
        categoryId, req.session.userId, amount, description, occurredAt
      );
      const expenseId = result.lastInsertRowid;

      // Ghi transaction
      q.run(
        `INSERT INTO transactions
           (type, amount, description, reference_type, reference_id, user_id, occurred_at)
         VALUES ('expense', ?, ?, 'expense', ?, ?, ?)`,
        amount, description, expenseId, req.session.userId, occurredAt
      );

      return expenseId;
    });

    addExpense();
    res.flash('success', 'Đã ghi nhận khoản chi.');
  } catch (err) {
    res.flash('error', err.message || 'Có lỗi khi ghi nhận khoản chi.');
  }

  res.redirect('/finance');
});

// ─────────────────────────────────────────────
// GET /finance/categories — danh mục chi phí
// ─────────────────────────────────────────────
router.get('/categories', requireAdmin, (req, res) => {
  const categories = q.all(
    `SELECT ec.*,
            COUNT(e.id) AS expense_count
     FROM expense_categories ec
     LEFT JOIN expenses e ON e.category_id = ec.id
     GROUP BY ec.id
     ORDER BY ec.name`
  );

  res.render('finance/categories.html', { categories });
});

// ─────────────────────────────────────────────
// POST /finance/categories — thêm danh mục
// ─────────────────────────────────────────────
router.post('/categories', requireAdmin, (req, res) => {
  const name        = (req.body.name || '').trim();
  const description = (req.body.description || '').trim();

  if (!name) {
    res.flash('error', 'Vui lòng nhập tên danh mục.');
    return res.redirect('/finance/categories');
  }

  try {
    q.run(
      `INSERT INTO expense_categories (name, description) VALUES (?, ?)`,
      name, description || null
    );
    res.flash('success', `Đã thêm danh mục "${name}".`);
  } catch (err) {
    res.flash('error', 'Danh mục đã tồn tại hoặc có lỗi xảy ra.');
  }

  res.redirect('/finance/categories');
});

// ─────────────────────────────────────────────
// POST /finance/categories/:id/delete — ẩn danh mục
// ─────────────────────────────────────────────
router.post('/categories/:id/delete', requireAdmin, (req, res) => {
  const catId = parseInt(req.params.id);

  // Kiểm tra còn khoản chi không
  const usageRow = q.get(
    `SELECT COUNT(*) AS cnt FROM expenses WHERE category_id = ?`,
    catId
  );
  if (usageRow && usageRow.cnt > 0) {
    res.flash(
      'error',
      `Danh mục đang được dùng bởi ${usageRow.cnt} khoản chi, không thể xóa.`
    );
    return res.redirect('/finance/categories');
  }

  q.run(
    `UPDATE expense_categories SET is_active = 0 WHERE id = ?`,
    catId
  );

  res.flash('success', 'Đã ẩn danh mục.');
  res.redirect('/finance/categories');
});

module.exports = router;
